### DISCUSSION TITLE
provide a high level overview of the underlying optimization methods used

### DISCUSSION THREAD

**Discussion by jerabaul29** [do-mpc/do-mpc#310] [Q&A] [answered]
Many thanks for an awesome project :) .

The documentation is very helpful for getting started with the framework and using it. However, I have a hard time finding information about the underlying optimization methods used. Is this documented somewhere (if so, sorry I missed it - if I did not "grossly missed it", may it be worth to make it more visible?)? If not, do you think there could be a way to provide a high level overview / point to resources? Based on the credits you provide, I guess https://web.casadi.org/docs/ and https://coin-or.github.io/Ipopt/ ; is there a way to give a high level overview of how these are used, or is it too complex to be possible?

**Comment by 4flixt** [comment, marked as answer]
Hi @jerabaul29,

thank you for your question. The tools that  youcited are what enable do-mpc. I think the best way to learn more about these methods is to [read this paper on IPOPT](https://drops.dagstuhl.de/volltexte/2009/2089/pdf/09061.WaechterAndreas.Paper.2089.pdf).
IPOPT solves the MPC optimization problem (in most cases). 
CasADi is used to supply IPOPT with the required first and second order derivatives of cost, constraints etc. and computes these through automatic differentiation. I suggest you read a bit more about automatic differentiation as well. Notice that this essentially the same as backpropagation (used for neural network training if you are familiar with that).

Best,
4flixt

**Reply by jerabaul29** [reply]
Many thanks for your kind answer and your high level explanations / pointers :) . I will look into this further :) .
